app/models/repository.rb

```diff
@@ -104,12 +104,31 @@ def self.add_repo(event, namespace, repo, tag)
       return
     end
 
-    digest = event.try(:[], "target").try(:[], "digest")
-    tag = repository.tags.create(name: tag, author: actor, digest: digest)
+    # And store the tag and its activity.
+    id, digest = Repository.id_and_digest_from_event(event, repo)
+    tag = repository.tags.create(name: tag, author: actor, digest: digest, image_id: id)
     repository.create_activity(:push, owner: actor, recipient: tag)
     repository
   end
 
+  # Fetch the image ID and the manifest digest from the given event.
+  def self.id_and_digest_from_event(event, repo)
+    digest = event.try(:[], "target").try(:[], "digest")
+
+    if digest.blank?
+      id = ""
+    else
+      begin
+        id, = Registry.get.client.manifest(repo, digest)
+      rescue StandardError => e
+        logger.warn "Could not fetch manifest for '#{repo}' with digest '#{digest}': " + e.message
+        id = ""
+      end
+    end
+
+    [id, digest]
+  end
+
```

Comment:
This is just personal taste, it can be written in that way:

```ruby
def self.id_and_digest_from_event(event, repo)
  digest = event.try(:[], "target").try(:[], "digest")
  id = ""
  unless digest.blank?
    begin
      id, = Registry.get.client.manifest(repo, digest)
    rescue StandardError => e
      logger.warn "Could not fetch manifest for '#{repo}' with digest '#{digest}': " + e.message
    end
  end

  [id, digest]
end
```